kano_profile/challenge_progress_tracker.py

```diff
@@ -19,12 +22,46 @@ class ChallengeProgressTracker(object):
     COMPLETED = 'completed'
     ABANDONED = 'abandoned'
 
+    _singleton_instance = None
+
+    @staticmethod
+    def get_instance():
+        if not ChallengeProgressTracker._singleton_instance:
+            ChallengeProgressTracker()
+
+        return ChallengeProgressTracker._singleton_instance
+
     def __init__(self):
+        if ChallengeProgressTracker._singleton_instance:
+            raise Exception('This class is a singleton!')
+        else:
+            ChallengeProgressTracker._singleton_instance = self
+
+        app_name = get_program_name()
+
+        self._app_name = app_name
         self._tracking_mode = None
         self._chal_group = None
         self._chal_no = None
         self._chal_outcome = None
-        self._app_session_id = str(uuid4())
+        self._app_session_id = get_session_unique_id(app_name, getpid())
```

Comment:
Would it be appropriate to log a warning if this doesn't return anything (i.e., the session tracker wasn't initialised yet). Or do you see use cases that won't be tied to a session?